File: scripts/analysis.R
library(sparklyr)
library(dplyr)

sc <- spark_connect(master = "yarn-client")

get_file <- function(year) { spark_read_parquet(sc, paste0("brfss",year), paste0("brfss/",year,".parquet")) %>% na.omit() %>% mutate(year=year)}
                                                
brfss_2011 <- get_file("2011")
brfss_2012 <- get_file("2012")
brfss_2013 <- get_file("2013")
brfss_2014 <- get_file("2014")
brfss_2015 <- get_file("2015")
#brfss <- rbind(brfss_2011, brfss_2012, brfss_2013, brfss_2014, brfss_2015) %>%
brfss <- brfss_2011 %>%
# Encode the values as strings so that the prediction system will perform one-hot encoding with no further effort
filter(CASTHM1 %in% c(1,2) ) %>%
mutate(ingoodhealth = if_else(RFHLTH == 1, 'yes', if_else(RFHLTH == 2, 'no', 'unknown'))) %>%
mutate(hascoverage = if_else(HCVU651 == 1, 'yes', if_else(HCVU651 == 2, 'no', 'unknown'))) %>%
mutate(hasheartproblem = if_else(MICHD == 1, 'yes', if_else(MICHD == 2, 'no', 'unknown'))) %>%
mutate(hasasthma = if_else(CASTHM1 == 1, 0, 1)) %>%
mutate(hasarthritis = if_else(DRDXAR1 == 1, 'yes', if_else(DRDXAR1 == 2, 'no', 'unknown'))) %>%
mutate(hadheartattack = if_else(CVDINFR4 == 1, 'yes', if_else(CVDINFR4 == 2, 'no', 'unknown'))) %>%
mutate(hadangina = if_else(CVDCRHD4 == 1, 'yes', if_else(CVDCRHD4 == 2, 'no', 'unknown'))) %>%
mutate(racegroup = if_else(RACEGR3 == 1, 'whitenonhispanic', if_else(RACEGR3 == 2, 'blacknonhispanic', if_else(RACEGR3 == 3, 'othernonhispanic', if_else(RACEGR3 == 4, 'multiracialnonhispanic', if_else(RACEGR3 == 5, 'hispanic', 'unknown' )))))) %>%
mutate(agegroup = if_else(AGE_G == 1, '18', if_else(AGE_G == 2, '25', if_else(AGE_G == 3, '35', if_else(AGE_G == 4, '45', if_else(AGE_G == 5, '55', if_else(AGE_G == 6, '65', 'unknown' ))))))) %>%
mutate(bmicategory = if_else(BMI5CAT == 1, 'underweight', if_else(BMI5CAT == 2, 'normalweight', if_else(BMI5CAT == 3, 'overweight', if_else(BMI5CAT == 4, 'obese',  'unknown' ))))) %>%
mutate(smoker = if_else(SMOKER3 == 1, 'everyday', if_else(SMOKER3 == 2, 'somedays', if_else(SMOKER3 == 3, 'former', if_else(SMOKER3 == 4, 'never', 'unknown' )))))




response <- 'hasasthma'

features <- c(
'ingoodhealth',
'hascoverage',
'hasheartproblem',
'hasarthritis',
'racegroup',
'agegroup',
'bmicategory',
'smoker',
'hadheartattack',
'hadangina')

# We will use a simple function to evaluate the various models given the test data.
model_eval <- function(model, test) {
  prediction <- sdf_predict(model, test)
#  accuracy <- ml_classification_eval(prediction, 'hasasthma', 'prediction', metric='accuracy')
    accuracy <- ml_binary_classification_eval(prediction, 'hasasthma', 'prediction')

  print(paste0("accuracy = ",round(accuracy,2)))
  truth <- select(prediction,hasasthma) %>% collect
  p <- select(prediction, prediction)  %>% collect
  table(p$prediction,truth$hasasthma)
}


# We'll partition the data set 75%/25% between training and testing sets
partitions <- sdf_partition(brfss, training=0.75, test=0.25, seed=1234567)
sdf_register(partitions, c("brfss_training", "brfss_testing"))
brfss_training <- tbl(sc,"brfss_training")
brfss_testing <- tbl(sc, "brfss_testing")
        
# We then evaluate multiple different models
model_eval(ml_logistic_regression(brfss_training, response=response, features=features), brfss_testing) 
model_eval(ml_decision_tree(brfss_training, response=response, features=features, type='classification'), brfss_testing)
model_eval(ml_gradient_boosted_trees(brfss_training, response=response, features=features, type='classification'), brfss_testing)
model_eval(ml_random_forest(brfss_training, response=response, features=features, type='classification'), brfss_testing)

# All of these models have the same accuracy of approx 91% - and this simply reflects the underlying 
# data skew by predicting 'no asthma' for every example!


# We will reconstruct the dataset and use undersampling to try to get a more nuanced model
brfss_no_asthma <- brfss %>% filter(hasasthma==0) %>% sdf_sample(0.4, replacement=TRUE, seed=12345)
brfss_with_asthma <- brfss %>% filter(hasasthma==1)
brfss <- rbind(brfss_no_asthma, brfss_with_asthma)

partitions <- sdf_partition(brfss, training=0.75, test=0.25, seed=1234567)
sdf_register(partitions, c("brfss_training", "brfss_testing"))
brfss_training <- tbl(sc,"brfss_training")
brfss_testing <- tbl(sc, "brfss_testing")

# We'll reevaluate the models
model_eval(ml_logistic_regression(brfss_training, response=response, features=features), brfss_testing) 
model_eval(ml_decision_tree(brfss_training, response=response, features=features, type='classification'), brfss_testing)
model_eval(ml_gradient_boosted_trees(brfss_training, response=response, features=features, type='classification'), brfss_testing)
model_eval(ml_random_forest(brfss_training, response=response, features=features, type='classification'), brfss_testing)


# This gets only an accuracy of 80%, with high sensitivity (~99%), and low specifity (~4.5%)
# So it rarely marks a false positive, but will very frequently mark a false negative. 

File: scripts/analysis.scala
import org.apache.spark.ml.classification.LogisticRegression;
import org.apache.spark.sql.functions._;
import org.apache.spark.ml.feature._;
import org.apache.spark.ml._;

  
val brfss = spark.read.format("parquet").load("brfss/2011.parquet");
  
//Drop the null valued rows
val brfss_no_nulls = brfss.na.drop().filter("CASTHM1 in (1,2)");
  


// Function to convert the 1, 2, to a yes/no/otherwise string  
def yes_no(arg : Int) : String = { 
    arg match {
      case 1 => "Yes";
      case 2 => "No";
      case _ => "unknown"

    }
}
//Function to encode the race group
def code_racegroup(racegroup: Int) : String = { racegroup match { case 1 => "whitenonhispanic"; case 2 => "blacknonhispanic"; case 3 => "othernonhispanic"; case 4 => "multiracialnonhispanic"; case 5 => "hispanic"; case _ => "unknown" } }
def code_agegroup(agegroup: Int) : String = { agegroup match { case 1 => "18+"; case 2 => "25+"; case 3 => "35+"; case 4 => "45+"; case 5 => "55+"; case 6 => "65+"; case _ => "unknown" } }
def code_bmicategory(bmicategory: Int) : String = { bmicategory match { case 1 => "underweight"; case 2 => "normalweight"; case 3 => "overweight"; case 4 => "obese";  case _ =>"unknown" } }
def code_smoker(smoker: Int) : String = { smoker match { case 1 => "everyday"; case 2 => "somedays"; case 3 => "former"; case 4 => "never"; case _ => "unknown" } }



val yes_no_udf = udf(yes_no _);
val code_racegroup_udf = udf(code_racegroup _);
val code_agegroup_udf = udf(code_agegroup _);
val code_bmicategory_udf = udf(code_bmicategory _);
val code_smoker_udf = udf(code_smoker _);
  
  
val brfss_labeled = brfss_no_nulls.withColumn("label", 'CASTHM1 - 1)
val brfss_strings = brfss_labeled.
  withColumn("ingoodhealth", yes_no_udf('RFHLTH)).
  withColumn("hascoverage", yes_no_udf('HCVU651)).
  withColumn("hasheartproblem", yes_no_udf('MICHD)).
  withColumn("hasarthritis", yes_no_udf('DRDXAR1)).
  withColumn("hadheartattack", yes_no_udf('CVDINFR4)).
  withColumn("hadangina", yes_no_udf('CVDCRHD4)).
  withColumn("racegroup", code_racegroup_udf('RACEGR3)).
  withColumn("agegroup", code_agegroup_udf('AGE_G)).
  withColumn("bmicategory", code_bmicategory_udf('BMI5CAT)).
  withColumn("smoker", code_smoker_udf('SMOKER3)) ;                                                                 
                                   

val ingoodhealth_idxr = new StringIndexer().setInputCol("ingoodhealth").setOutputCol("ingoodhealth_idx");
val ingoodhealth_ohr = new OneHotEncoder().setInputCol("ingoodhealth_idx").setOutputCol("ingoodhealth_ohe");
val hascoverage_idxr = new StringIndexer().setInputCol("hascoverage").setOutputCol("hascoverage_idx");
val hascoverage_ohr = new OneHotEncoder().setInputCol("hascoverage_idx").setOutputCol("hascoverage_ohe");
val hasheartproblem_idxr = new StringIndexer().setInputCol("hasheartproblem").setOutputCol("hasheartproblem_idx");
val hasheartproblem_ohr = new OneHotEncoder().setInputCol("hasheartproblem_idx").setOutputCol("hasheartproblem_ohe");
val hasarthritis_idxr = new StringIndexer().setInputCol("hasarthritis").setOutputCol("hasarthritis_idx");
val hasarthritis_ohr = new OneHotEncoder().setInputCol("hasarthritis_idx").setOutputCol("hasarthritis_ohe");
val hadheartattack_idxr = new StringIndexer().setInputCol("hadheartattack").setOutputCol("hadheartattack_idx");
val hadheartattack_ohr = new OneHotEncoder().setInputCol("hadheartattack_idx").setOutputCol("hadheartattack_ohe");
val hadangina_idxr = new StringIndexer().setInputCol("hadangina").setOutputCol("hadangina_idx");
val hadangina_ohr = new OneHotEncoder().setInputCol("hadangina_idx").setOutputCol("hadangina_ohe");
val racegroup_idxr = new StringIndexer().setInputCol("racegroup").setOutputCol("racegroup_idx");
val racegroup_ohr = new OneHotEncoder().setInputCol("racegroup_idx").setOutputCol("racegroup_ohe");
val agegroup_idxr = new StringIndexer().setInputCol("agegroup").setOutputCol("agegroup_idx");
val agegroup_ohr = new OneHotEncoder().setInputCol("agegroup_idx").setOutputCol("agegroup_ohe");
val bmicategory_idxr = new StringIndexer().setInputCol("bmicategory").setOutputCol("bmicategory_idx");
val bmicategory_ohr = new OneHotEncoder().setInputCol("bmicategory_idx").setOutputCol("bmicategory_ohe");
val smoker_idxr = new StringIndexer().setInputCol("smoker").setOutputCol("smoker_idx");
val smoker_ohr = new OneHotEncoder().setInputCol("smoker_idx").setOutputCol("smoker_ohe");
                                       

                                       
 val assembler = new VectorAssembler().setInputCols(
   Array("ingoodhealth_ohe",
        "hascoverage_ohe",
        "hasheartproblem_ohe",
        "hasarthritis_ohe",
        "hadheartattack_ohe",
        "hadangina_ohe",
        "racegroup_ohe",
        "agegroup_ohe",
        "bmicategory_ohe",
        "smoker_ohe")).
   setOutputCol("features"); 
                                       
val lr = new LogisticRegression().setMaxIter(10).setRegParam(0.3).setElasticNetParam(0.8);

val transform_pipeline = new Pipeline().setStages(
  Array(ingoodhealth_idxr,
        ingoodhealth_ohr,
        hascoverage_idxr,
        hascoverage_ohr,
        hasheartproblem_idxr,
        hasheartproblem_ohr,
        hasarthritis_idxr,
        hasarthritis_ohr,
        hadheartattack_idxr,
        hadheartattack_ohr,
        hadangina_idxr,
        hadangina_ohr,
        racegroup_idxr,
        racegroup_ohr,
        agegroup_idxr,
        agegroup_ohr,
        bmicategory_idxr,
        bmicategory_ohr,
        smoker_idxr,
        smoker_ohr,
        assembler));
                                       
val brfss_transformer = transform_pipeline.fit(brfss_strings);
val brfss_transformed = brfss_transformer.transform(brfss_strings);
                                       

                                       
                            
val splits = brfss_transformed.randomSplit(Array(0.75, 0.25),12345) ;                                                                         
val training = splits(0).cache();
val test = splits(1).cache();
  
                                    

  
  // Fit the model
val lrModel = lr.fit(training);
// make a prediction
val p = lrModel.transform(test)
                                       
val tp = p.filter("label == 1 and prediction ==1").count;
val fp = p.filter("label == 0 and prediction ==1").count;
val tn = p.filter("label == 0 and prediction ==0").count;
val fn = p.filter("label == 1 and prediction == 0").count;
                                       
spark.createDataFrame(Seq(
 ("Predicted +", tp, fp),
  ("Predicted -",fn, tn)
)).toDF("","Actual +", " Actual -").show()
                                       
                                      

